### PR TITLE
Read location group

### PR DESCRIPTION
**Pull request by LorenzoCianciaruso**

### CODE REVIEW

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 1]
```diff
@@ -111,14 +138,20 @@ private boolean isValidAndUnexpired(Optional<ClusterCacheEntry> entry) {
...
     }
 
-    private MessageResults readMessages(List<String> types, long start, long startOffset, List<Long> clusterIds, Connection connection) throws SQLException {
+    private MessageResults readMessages(
+        List<String> types,
+        long start,
+        long startOffset,
+        List<Long> clusterIds,
+        List<Long> locationGroups,
+        Connection connection) throws SQLException {
```
Comment (adam-yorwerth) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:147]: formatting

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 2]
```diff
@@ -738,19 +814,23 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
-                "INSERT INTO EVENTS(msg_offset, msg_key, content_type, type, created_utc, data, event_size, cluster_id) VALUES(?,?,?,?,?,?,?,?);",
-                msg.offset, msg.key, msg.contentType, msg.type, time, msg.data, messageSize, clusterId
-            )
-        }
+    void insert(
+        Message msg,
+        Long clusterId,
+        int messageSize=0,
+        Timestamp time = Timestamp.valueOf(msg.created.toLocalDateTime()),
+        Long locationGroup = null) {
```
Comment (sebastianlavender) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:822]: Formatting looks a little off here

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 3]
```diff
@@ -727,7 +734,76 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
+
+        when: "messages are read"
+        def messageResults = storage.read([], 0L, "locationUuid")
+
+        then: "only the message belonging to the location cluster is returned"
+        messageResults.messages.size() == 1
+        messageResults.messages.get(0).offset == 1
+    }
+
+    def "messages with location group are not read if location is not pressent in location groups table"() {
```
Comment (ashish-sharma09) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:754]: type in present

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 4]
```diff
@@ -727,7 +734,76 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
         where:
         types << [ [], ["type1"] ]
     }
-    
+
+    def "messages with location group are not read if location is not part of any groups"() {
```
Comment (ashish-sharma09) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:738]: Probably  something like - if "given (or requested)" location is not part of any groups.

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 5]
```diff
@@ -727,7 +734,76 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
+        insert(message(4, "type1", "D", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"), 3L, 0, Timestamp.valueOf(time.toLocalDateTime()), 3L)
+        insert(message(5, "type1", "E", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"), 1L, 0, Timestamp.valueOf(time.toLocalDateTime()), 3L)
+        insert(message(6, "type1", "F", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"), 3L, 0, Timestamp.valueOf(time.toLocalDateTime()), 4L)
+
+        when: "messages are read"
+        def messageResults = storage.read([], 0L, locationUuid)
+
+        then: "only messages for the relevant cluster and group are returned"
+        messageResults.messages.size() == 3
+        messageResults.messages*.offset*.intValue() == [3, 4, 6]
```
Comment (sebastianlavender) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:803]: I think the first * is not needed here as the messages are already a list so don't need collecting

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 6]
```diff
@@ -727,7 +734,76 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
     }
-    
+
+    def "messages with location group are not read if location is not part of any groups"() {
+        given: "2 messages, one of which with a location group that the location does not belong to"
+        insert(message(1, "type1", "A", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"))
+        insert(message(2, "type1", "A", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"), 1L, 0, Timestamp.valueOf(time.toLocalDateTime()), 2L)
+
+        and: "location does not belong to the group"
+        insertLocationGroupFor("locationUuid", [])
```
Comment (ashish-sharma09) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:744]: In real situation there should not be a location with empty group, probably not having any data in locationgroups table is less confusing

Comment (LorenzoCianciaruso) [reply pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:744]: I think it depends if we are gonna remove the location groups if there aren't any messages for those groups in the event table and how we implement it (do we remove the row if no location groups, or do we just remove the id and leave an empty list if it was the only group) . Really the test just makes sure that we handle it correctly even if the list is empty

**Review by sebastianlavender — APPROVED** [review]
LGTM

**Review by ashish-sharma09 — APPROVED** [review]
LGTM

**Pull request merged**